src/main/java/core/components/GridBoard.java

```diff
@@ -156,7 +149,7 @@ public T getElement(int x, int y) {
      *
      * @return - 2D grid.
      */
-    public T[][] getGridValues() {
+    public Component[][] getGridValues() {
```

Comment:
getElement() and setElement() use the more natural T as parameter/return value. Why not do this here too (as indeed was the case before the change) and return T[][]? Admittedly there is a need to cast a created array to T[][]...but we can't avoid casts to T at some point in the code.

Reply:
I would have liked that too. I prefer to stay away from reflection, which would be needed for casting the 2D array type (can't just use normal casting with array types). Since the type is known when the class is used, it's easier and faster to cast on the other side (though admittedly more work). Users can also use the get/set element methods which do cast it, so I thought it might be fine to not complicate it here? 

If not, do you know if java 8 streams or reflection would be faster? Both would be options here to return the T[][] type.

Reply:
I've refreshed my understanding of T[] vs T casts - and in particular why casting to T[] when we know the actual array is Object[] is a poor idea (while casting to T is fine as we are confident that the actual content is a T).

While there are methods to deal with this - they might just confuse people. So I agree with the proposed solution.

Except...flattenGrid needs to be amended to return Component[] instead of T[] for consistency ;-)

Reply:
Fair point! Updated :)